src/runtime/mutex.ts

```diff
@@ -0,0 +1,65 @@
+/**
+ * Running a Releaser releases the acquired mutex.
+ */
+export interface Releaser {
+  (): void;
+}
+
+/**
+ * A simmple Mutex to gate access to critical async code
+ * sections that should not execute concurrently.
+ *
+ * Sample usage:
+ * 
+ * ```
+ *   class SampleClass {
+ *     private readonly mutex = new Mutex();
+ *
+ *     async instantiate() {
+ *       const release = await mutex.acquire();
+ *       try {
+ *         // Protected section with async execution.
+ *       } finally {
+ *         release();
+ *       }
+ *     }
+ *   }
+ */
+
+export class Mutex {
+  private next: Promise<void> = Promise.resolve();
+  private depth = 0; // tracks the number of blocked executions on this lock.
+
+  /**
+   * @return true if the mutex is already acquired.
+   */
+  get locked(): boolean {
```

Comment:
YAGNI?

Reply:
I already used it to log stacktraces before acquiring